Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -547,6 +548,81 @@ extension SwiftLanguageServer {
     // FIXME: cancellation
     _ = handle
   }
+
+  func foldingRange(_ req: Request<FoldingRangeRequest>) {
+
+    guard let snapshot = documentManager.latestSnapshot(req.params.textDocument.url) else {
+      log("failed to find snapshot for url \(req.params.textDocument.url)")
+      req.reply(nil)
+      return
+    }
+
+    let skreq = SKRequestDictionary(sourcekitd: sourcekitd)
+    skreq[keys.request] = requests.editor_open
+    skreq[keys.name] = "FoldingRanges:" + snapshot.document.url.path
+    skreq[keys.sourcetext] = snapshot.text
+    skreq[keys.syntactic_only] = 1
+
+    let handle = sourcekitd.send(skreq) { [weak self] result in
+      guard let self = self else { return }
+      guard let dict = result.success else {
+        req.reply(.failure(result.failure!))
+        return
+      }
+
+      guard let syntaxMap: SKResponseArray = dict[self.keys.syntaxmap],
+            let substructure: SKResponseArray = dict[self.keys.substructure] else {
+        return req.reply([])
+      }
+
+      var ranges: [FoldingRange] = []
+
+      syntaxMap.forEach { _, value in
+        if let kind: sourcekitd_uid_t = value[self.keys.kind],
+           kind == self.values.syntaxtype_comment,
+           let offset: Int = value[self.keys.offset],
+           let start: Position = snapshot.positionOf(utf8Offset: offset),
+           let length: Int = value[self.keys.length],
+           // SourceKit marks the end of a comment as the first non-comment character
+           // after it, so we subtract one to get the real comment range.
+           let end: Position = snapshot.positionOf(utf8Offset: offset + length - 1) {
+          let range = FoldingRange(startLine: start.line,
```

Comment:
Style: When you have multiple lines of `if let` bindings and conditions, please put the final `{` on its own line. Otherwise it's hard to see where the condition ends and the following code begins (I read this wrong twice now).  There's a similar case below in the structure handling code.